### PR TITLE
Optimize docker image (re)builds

### PR DESCRIPTION
**Pull request by rzadp**
Some optimization mechanics for the docker images - very useful when changing the code and rebuilding the images.

## Dockerfiles

1. Move the `apk add` commands to top - so that this layer can be cached between code changes
2. Split the step of `yarn install`, so that the layer with `node_modules` can be cached between code changes

## Dockerignore

There are a lot of unnecessary stuff in the docker build context:

```bash
./README.md
./server_injected.Dockerfile
./tsconfig.json
./jest.config.js
./env.server.config.json
./.eslintignore
./env.bot.config.json
./.prettierignore
./nodemon.json
./bot_injected.Dockerfile
./.eslintrc.cjs
./.idea
./.idea/workspace.xml
./.idea/modules.xml
./.idea/.gitignore
./.idea/optimize-faucet.iml
./.idea/vcs.xml
./.gitlab-ci.yml
./yarn.lock
./src
./src/faucetConfig.ts
./src/logger.ts
./src/utils.ts
./src/server
./src/server/config.ts
./src/server/constants.ts
./src/server/routes
./src/server/routes/healthcheck.ts
./src/server/routes/metrics.ts
./src/server/routes/actions.ts
./src/server/utils.ts
./src/server/router.ts
./src/server/utils.spec.ts
./src/server/services
./src/server/services/DripRequestHandler.ts
./src/server/services/Recaptcha.ts
./src/server/services/ActionStorage.spec.ts
./src/server/services/Actions.ts
./src/server/services/Recaptcha.spec.ts
./src/server/services/ErrorCounter.ts
./src/server/services/ActionStorage.ts
./src/server/services/DripRequestHandler.spec.ts
./src/server/polkadotApi.ts
./src/server/index.ts
./src/faucet.e2e.ts
./src/guards.ts
./src/utils.spec.ts
./src/bot
./src/bot/index.ts
./src/types.ts
./CODEOWNERS
./docker
./docker/docker-compose.rococo.yml
./docker/docker-compose.westend.yml
./docker/docker-compose.versi.yml
./docker/docker-compose.yml
./docker/docker-compose.rococo-external.yml
./e2e
./e2e/matrix_data
./e2e/matrix_data/localhost.log.config
./e2e/matrix_data/homeserver.yaml
./e2e/matrix_data/.gitignore
./e2e/matrix_data/localhost.signing.mockkey
./e2e/wait_until.sh
./e2e/docker-compose.yml
./e2e/bootstrap.sh
./.editorconfig
./kubernetes
./kubernetes/rococo-values.yaml
./kubernetes/rococo-external-values.yaml
./kubernetes/westend-values.yaml
./.gitignore
./package.json
./.prettierrc.cjs
./.dockerignore
./Dockerfile.build-context
./.github
./.github/workflows
./.github/workflows/E2E.yml
./jest.e2e.config.js
./.env.example
./.git
./.git/hooks
./.git/hooks/pre-applypatch.sample
./.git/hooks/applypatch-msg.sample
./.git/hooks/pre-rebase.sample
./.git/hooks/fsmonitor-watchman.sample
./.git/hooks/pre-merge-commit.sample
./.git/hooks/pre-commit.sample
./.git/hooks/push-to-checkout.sample
./.git/hooks/prepare-commit-msg.sample
./.git/hooks/post-update.sample
./.git/hooks/pre-push.sample
./.git/hooks/commit-msg.sample
./.git/hooks/pre-receive.sample
./.git/hooks/update.sample
./.git/description
./.git/info
./.git/info/exclude
./.git/HEAD
./.git/index
./.git/objects
./.git/objects/info
./.git/objects/pack
./.git/objects/pack/pack-b1da18ccc15d6000e6f7f9b53066574597954d68.idx
./.git/objects/pack/pack-b1da18ccc15d6000e6f7f9b53066574597954d68.pack
./.git/refs
./.git/refs/heads
./.git/refs/heads/main
./.git/refs/heads/rzadp
./.git/refs/heads/rzadp/optimize-docker
./.git/refs/tags
./.git/refs/remotes
./.git/refs/remotes/origin
./.git/refs/remotes/origin/HEAD
./.git/logs
./.git/logs/HEAD
./.git/logs/refs
./.git/logs/refs/heads
./.git/logs/refs/heads/main
./.git/logs/refs/heads/rzadp
./.git/logs/refs/heads/rzadp/optimize-docker
./.git/logs/refs/remotes
./.git/logs/refs/remotes/origin
./.git/logs/refs/remotes/origin/HEAD
./.git/config
./.git/packed-refs
./.git/FETCH_HEAD
```

The problem with this is that any change to any of those files, even if it is not related to typescript code, causes the `COPY . . => yarn build` layers to be invalidated.

I went with the extremist approach of excluding everything and then including what is needed - but I don't have a strong preference to this approach.

### CODE REVIEW

File: bot_injected.Dockerfile [hunk 1]
```diff
@@ -15,9 +19,12 @@ LABEL io.parity.image.authors="cicd-team@parity.io" \
     io.parity.image.created="${BUILD_DATE}"
 
 WORKDIR /bot
+
+COPY ./package.json .
+COPY ./yarn.lock .
```
Comment (Bullrich) [inline bot_injected.Dockerfile:24]: You can move this to a single line
```suggestion
COPY ./package.json ./yarn.lock .
```

File: server_injected.Dockerfile [hunk 2]
```diff
@@ -15,9 +19,13 @@ LABEL io.parity.image.authors="cicd-team@parity.io" \
     io.parity.image.created="${BUILD_DATE}"
 
 WORKDIR /backend
+
+COPY ./package.json .
+COPY ./yarn.lock .
```
Comment (Bullrich) [inline server_injected.Dockerfile:24]: Also here
```suggestion
COPY ./package.json ./yarn.lock .
```

File: server_injected.Dockerfile [hunk 3]
```diff
@@ -15,9 +19,13 @@ LABEL io.parity.image.authors="cicd-team@parity.io" \
...
+RUN yarn --network-concurrency 1 --frozen-lockfile
+
 COPY . .
-# uncomment to fix build on MacOS Apple Silicon chip
-# RUN apk add --no-cache python3 make g++
-RUN apk add git==2.30.6-r0
+RUN yarn prepare && yarn build
```
Comment (Bullrich) [inline server_injected.Dockerfile:28]: Wouldn't be better to change the `prepare` script to `prebuild` so it's done automatically?

**Review by Bullrich — APPROVED** [review]
I find this PR very positive! It's good to remove all this unnecessary files.

If you want to optimize the image even more, we can use a two stage docker image to only execute the build.

I did that on [the GitHub-Issue-Sync Dockerfile](https://github.com/paritytech/github-issue-sync/blob/action-migration/Dockerfile) if it works as a reference.

More info: https://docs.docker.com/build/building/multi-stage/

File: bot_injected.Dockerfile (hunk 1 above)
Comment (mutantcornholio) [reply bot_injected.Dockerfile:24]: +1, otherwise it'd create a separate layer for each copy command

Comment (rzadp) [reply bot_injected.Dockerfile]: Done ✅

File: server_injected.Dockerfile (hunk 2 above)
Comment (rzadp) [reply server_injected.Dockerfile]: Done as well ✅

File: server_injected.Dockerfile (hunk 3 above)
Comment (rzadp) [reply server_injected.Dockerfile]: Sounds good, changed ✅

**Comment by rzadp** [comment]
> If you want to optimize the image even more, we can use a two stage docker image to only execute the build.

Great suggestion. We could also think about having one Dockerfile with a common stage, a stage for bot and a stage for server - since the existing two Dockerfiles are almost identical. Anyways I would leave it for separate PR in order to have smaller changes.

**Review by sergejparity — APPROVED** [review]

**Pull request merged**